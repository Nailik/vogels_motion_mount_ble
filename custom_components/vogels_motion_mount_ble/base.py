"""Base entity which all other entity platform classes can inherit.

As all entity types have a common set of properties, you can
create a base entity like this and inherit it in all your entity platforms.

This just makes your code more efficient and is totally optional.

See each entity platform (ie sensor.py, switch.py) for how this is inheritted
and what additional properties and methods you need to add for each entity type.

"""

import logging
from typing import Any

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN
from .coordinator import ExampleCoordinator

_LOGGER = logging.getLogger(__name__)


class ExampleBaseEntity(CoordinatorEntity):
    """Base Entity Class.

    This inherits a CoordinatorEntity class to register your entites to be updated
    by your DataUpdateCoordinator when async_update_data is called, either on the scheduled
    interval or by forcing an update.
    """

    coordinator: ExampleCoordinator

    # ----------------------------------------------------------------------------
    # Using attr_has_entity_name = True causes HA to name your entities with the
    # device name and entity name.  Ie if your name property of your entity is
    # Voltage and this entity belongs to a device, Lounge Socket, this will name
    # your entity to be sensor.lounge_socket_voltage
    #
    # It is highly recommended (by me) to use this to give a good name structure
    # to your entities.  However, totally optional.
    # ----------------------------------------------------------------------------
    _attr_has_entity_name = True

    def __init__(self, coordinator: ExampleCoordinator) -> None:
        """Initialise entity."""
        super().__init__(coordinator)

    @property
    def device_info(self) -> DeviceInfo:
        """Return device information."""

        # ----------------------------------------------------------------------------
        # Identifiers are what group entities into the same device.
        # If your device is created elsewhere, you can just specify the indentifiers
        # parameter to link an entity to a device.
        # If your device connects via another device, add via_device parameter with
        # the indentifiers of that device.
        #
        # Device identifiers should be unique, so use your integration name (DOMAIN)
        # and a device uuid, mac address or some other unique attribute.
        # ----------------------------------------------------------------------------
        return DeviceInfo(
            name=self.coordinator.name,
            manufacturer="Vogelsr",
            model="Motion Mount",
            identifiers={(DOMAIN, self.coordinator.mac)},
        )

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return self.coordinator.name

    @property
    def unique_id(self) -> str:
        """Return unique id."""

        # ----------------------------------------------------------------------------
        # All entities must have a unique id across your whole Home Assistant server -
        # and that also goes for anyone using your integration who may have many other
        # integrations loaded.
        #
        # Think carefully what you want this to be as changing it later will cause HA
        # to create new entities.
        #
        # It is recommended to have your integration name (DOMAIN), some unique id
        # from your device such as a UUID, MAC address etc (not IP address) and then
        # something unique to your entity (like name - as this would be unique on a
        # device)
        #
        # If in your situation you have some hub that connects to devices which then
        # you want to create multiple sensors for each device, you would do something
        # like.
        #
        # f"{DOMAIN}-{HUB_MAC_ADDRESS}-{DEVICE_UID}-{ENTITY_NAME}""
        #
        # This is even more important if your integration supports multiple instances.
        # ----------------------------------------------------------------------------
        return f"{DOMAIN}-{self.coordinator.mac}"
