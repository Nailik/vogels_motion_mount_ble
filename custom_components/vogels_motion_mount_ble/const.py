"""Constants for the Vogels Motion Mount BLE integration."""

DOMAIN = "vogels_motion_mount_ble"

SERVICE_UUID = "3e6fe65d-ed78-11e4-895e-00026fd5c52c"

CHAR_NAME_UUID = "c005fa37-0651-4800-b000-000000000000"
CHAR_SIZE_UUID = "c005fa2b-0651-4800-b000-000000000000"
CHAR_DISTANCE_UUID = "c005fa00-0651-4800-b000-000000000000"
CHAR_ROTATION_UUID = "c005fa01-0651-4800-b000-000000000000"
CHAR_AUTOMATIC_UUID = "c005fa02-0651-4800-b000-000000000000"
CHAR_FREEZE_UUID = "c005fa14-0651-4800-b000-000000000000"
CHAR_PRESET_UUIDS = [
    "c005fa0a-0651-4800-b000-000000000000",
    "c005fa0b-0651-4800-b000-000000000000",
    "c005fa0c-0651-4800-b000-000000000000",
    "c005fa0d-0651-4800-b000-000000000000",
    "c005fa0e-0651-4800-b000-000000000000",
    "c005fa0f-0651-4800-b000-000000000000",
    "c005fa10-0651-4800-b000-000000000000",
]
CHAR_PRESET_UUID = "c005fa2a-0651-4800-b000-000000000000"
