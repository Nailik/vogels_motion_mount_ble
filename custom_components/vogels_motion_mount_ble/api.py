"""API Placeholder.

You should create your api seperately and have it hosted on PYPI.  This is included here for the sole purpose
of making this example code executable.
"""

import asyncio
from dataclasses import dataclass, replace
import logging
from .const import CHAR_DISTANCE_UUID, CHAR_ROTATION_UUID, CHAR_PRESET_UUID
from bleak import BleakClient, BleakScanner

_LOGGER = logging.getLogger(__name__)

@dataclass
class VogelsMotionMountData:
    """Class to hold api data."""

    connected: bool = False
    distance: int | None = None
    rotation: int | None = None

class API:
    """Class for example API."""

    def __init__(self, mac: str, pin: str | None, coordinator) -> None:
        """Initialise."""
        self._mac = mac
        self._pin = pin
        self._coordinator = coordinator
        self._data = VogelsMotionMountData(connected=False)
        self._client: BleakClient | None = None
        self._disconnected_event = asyncio.Event()

    async def testConnect(self):
        """Test connection to the BLE device once using BleakClient."""
        try:
            _LOGGER.debug("start test connection")
            async with BleakClient(self._mac, timeout=120) as client:
                _LOGGER.debug("Connected to device: %s", client.is_connected)
        except Exception as err:
            _LOGGER.error("Failed to connect to device: %s", err)
            raise APIConnectionError("Error connecting to api.") from err


    async def perform_action(self):
        _LOGGER.exception("write_gatt_char")
        result = await self._client.write_gatt_char(CHAR_PRESET_UUID, bytes([1]), response=True)
        _LOGGER.exception("write_gatt_char result: %s", result)

    def update(self, **kwargs):
        """Update one or more fields, retaining others from existing data. Then notify the coordinator."""
        self._data = replace(self._data, **kwargs)
        self._coordinator.async_set_updated_data(self._data)

    def _handle_disconnect(self, _):
        _LOGGER.debug("handle Device disconnected!")
        self._disconnected_event.set()

    def _handle_distance_change(self, _, data):
        _LOGGER.debug("distance change %s", data)
        self.update(distance=int.from_bytes(data, "little"))

    def _handle_rotation_change(self, _, data):
        _LOGGER.debug("rotation change %s", data)
        self.update(rotation=int.from_bytes(data, "little"))

    # todo make it optional if the connection should be maintained or connect on command (when required to send command) or poll time
    async def maintainConnection(self):
        """Maintain connection to device."""
        while True:
            _LOGGER.debug("scanning for device %s", self._mac)
            device = await BleakScanner.find_device_by_address(self._mac)

            if device is None:
                _LOGGER.debug("no device %s found, wait then scan again", self._mac)
                await asyncio.sleep(5)
                continue

            try:
                _LOGGER.debug("connecting to device")
                async with BleakClient(device, disconnected_callback=self._handle_disconnect, timeout=120) as client:
                    self.connected = client.is_connected
                    self._client = client
                    _LOGGER.debug("maintain connected to device %s", self._client.is_connected)
                    self.update(connected = self._client.is_connected)

                    self._handle_distance_change(None, await client.read_gatt_char(CHAR_DISTANCE_UUID))
                    self._handle_rotation_change(None, await client.read_gatt_char(CHAR_ROTATION_UUID))

                    await client.start_notify(CHAR_DISTANCE_UUID, self._handle_distance_change)
                    await client.start_notify(CHAR_ROTATION_UUID, self._handle_rotation_change)

                    await self._disconnected_event.wait()
                    # reset event
                    self._disconnected_event.clear()
                    _LOGGER.debug("device disconnected %s", self.connected)
                    self.update(connected = client.is_connected)
                    await asyncio.sleep(1)
            except Exception:
                # catch bleak.exc.BleakError: No backend with an available connection slot that can reach address D9:13:5D:AB:3B:37 was found
                _LOGGER.exception("Exception while connecting/connected")
                await asyncio.sleep(5)


class APIConnectionError(Exception):
    """Exception class for connection error."""
