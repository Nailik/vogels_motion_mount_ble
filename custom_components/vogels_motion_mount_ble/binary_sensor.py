"""Interfaces with the Integration 101 Template api sensors."""

import logging

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from . import MyConfigEntry
from .base import ExampleBaseEntity
from .const import DOMAIN
from homeassistant.helpers import config_validation as cv
import voluptuous as vol
from homeassistant.helpers.typing import ConfigType
from homeassistant.core import HomeAssistant, ServiceCall, callback

_LOGGER = logging.getLogger(__name__)

async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: MyConfigEntry,
    async_add_entities: AddEntitiesCallback,
):
    """Set up the Binary Sensors."""
    coordinator = hass.data[DOMAIN][config_entry.entry_id]
    async_add_entities([ExampleBinarySensor(coordinator)])


class ExampleBinarySensor(ExampleBaseEntity, BinarySensorEntity):
    """Implementation of a sensor."""

    _attr_device_class = BinarySensorDeviceClass.CONNECTIVITY
    _attr_has_entity_name = True
    _attr_translation_key = "connection"
    _attr_should_poll = True

    @callback
    def _handle_coordinator_update(self) -> None:
        """Update sensor with latest data from coordinator."""
        _LOGGER.exception("_handle_coordinator_update binary sensor")
        # This method is called by your DataUpdateCoordinator when a successful update runs.
        self.async_write_ha_state()

    @property
    def name(self) -> str:
        """Return the name of the sensor."""
        return "connection"

    @property
    def unique_id(self) -> str:
        """Return unique id."""
        # All entities must have a unique id.  Think carefully what you want this to be as
        # changing it later will cause HA to create new entities.
        return f"{DOMAIN}-{self.coordinator.mac}-connected"

    @property
    def is_on(self) -> bool:
        """Return if the binary sensor is on."""
        _LOGGER.exception("is_on binary sensor")
        if not self.coordinator.data:
            _LOGGER.error("No data available in coordinator")
            return None
        # This needs to enumerate to true or false
        return self.coordinator.data.connected

