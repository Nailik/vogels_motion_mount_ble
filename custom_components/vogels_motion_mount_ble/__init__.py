"""Integration for a Vogels Motion Mount via BLE."""

from __future__ import annotations

import logging
from homeassistant.const import __version__ as ha_version
from packaging import version
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import (
    ConfigEntryAuthFailed,
    ConfigEntryError,
    ConfigEntryNotReady,
)

from .api import APIAuthenticationError, APIConnectionDeviceNotFoundError
from .coordinator import VogelsMotionMountBleCoordinator
from .services import async_setup_services

_LOGGER = logging.getLogger(__name__)

PLATFORMS: list[Platform] = [
    Platform.BINARY_SENSOR,
    Platform.BUTTON,
    Platform.NUMBER,
    Platform.SELECT,
    Platform.SENSOR,
    Platform.SWITCH,
    Platform.TEXT,
]

type VogelsMotionMountBleConfigEntry = ConfigEntry[VogelsMotionMountBleCoordinator]

# for this version read_only was added to config selectors
MIN_HA_VERSION = "2025.6.0"

if version.parse(ha_version) < version.parse(MIN_HA_VERSION):
    raise RuntimeError(
        f"Vogels Motion Mount BLE requires Home Assistant {MIN_HA_VERSION}+"
    )


async def async_setup(
    hass: HomeAssistant, entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Set up Vogels Motion Mount integration services."""
    _LOGGER.debug("async_setup called with config_entry: %s", entry)
    async_setup_services(hass)
    return True


async def async_setup_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Set up Vogels Motion Mount Integration from a config entry."""
    _LOGGER.debug("async_setup_entry called with config_entry: %s", config_entry)

    # Registers update listener to update config entry when options are updated.
    unsub_update_listener = config_entry.add_update_listener(async_reload_entry)

    # Initialise the coordinator that manages data updates from your api.
    coordinator = VogelsMotionMountBleCoordinator(
        hass=hass,
        config_entry=config_entry,
        unsub_update_listener=unsub_update_listener,
    )

    try:
        await coordinator.api.refresh_data()
    except APIConnectionDeviceNotFoundError as err:
        raise ConfigEntryNotReady("error_device_not_found") from err
    except APIAuthenticationError as err:
        raise ConfigEntryAuthFailed("error_invalid_athentication") from err
    except Exception as err:
        raise ConfigEntryError(
            translation_key=f"Something went wrong {err}",
            translation_placeholders={"error": err},
        ) from err

    config_entry.runtime_data = coordinator
    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
    return True


async def async_reload_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> None:
    """Reload config entry."""
    _LOGGER.debug(
        "async_reload_entry async_reload with pin %s", config_entry.data["conf_pin"]
    )
    await hass.config_entries.async_reload(config_entry.entry_id)


async def async_unload_entry(
    hass: HomeAssistant, config_entry: VogelsMotionMountBleConfigEntry
) -> bool:
    """Unload a config entry."""
    _LOGGER.debug("async_unload_entry")
    if unload_ok := await hass.config_entries.async_unload_platforms(
        config_entry, PLATFORMS
    ):
        _LOGGER.debug("async_unload_entry pop")
        coordinator: VogelsMotionMountBleCoordinator = config_entry.runtime_data
        await coordinator.unload()

    return unload_ok
