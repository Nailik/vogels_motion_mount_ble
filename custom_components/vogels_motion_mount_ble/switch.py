"""Button entities to define actions for Vogels Motion Mount BLE entities."""

import logging

from homeassistant.components.switch import SwitchEntity
from homeassistant.core import HomeAssistant, ServiceCall
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from . import VogelsMotionMountBleConfigEntry
from .api import VogelsMotionMountPinSettings
from .base import VogelsMotionMountBleBaseEntity, VogelsMotionMountBlePresetBaseEntity
from .const import (
    DOMAIN,
    HA_SERVICE_SET_MULTI_PIN_FEATURES,
    HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_PRESET_ID,
    HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_NAME_ID,
    HA_SERVICE_MULTI_PIN_FEATURE_DISABLE_CHANNEL_ID,
    HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_TV_ON_OFF_DETECTION_ID,
    HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_DEFAULT_PRESET_ID,
    HA_SERVICE_MULTI_PIN_FEATURE_START_CALIBRATION_ID,
    )
from .coordinator import VogelsMotionMountBleCoordinator
from .utils import get_coordinator

_LOGGER = logging.getLogger(__name__)


async def _set_multi_pin_features(call: ServiceCall) -> None:
    _LOGGER.debug("Set multi pin features change presets service called with data: %s", call.data)
    await get_coordinator(call).api.set_multi_pin_features(
        change_presets = call.data.get(HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_PRESET_ID),
        change_name = call.data.get(HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_NAME_ID),
        disable_channel = call.data.get(HA_SERVICE_MULTI_PIN_FEATURE_DISABLE_CHANNEL_ID),
        change_tv_on_off_detection = call.data.get(HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_TV_ON_OFF_DETECTION_ID),
        change_default_position = call.data.get(HA_SERVICE_MULTI_PIN_FEATURE_CHANGE_DEFAULT_PRESET_ID),
        start_calibration = call.data.get(HA_SERVICE_MULTI_PIN_FEATURE_START_CALIBRATION_ID),
    )

async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: VogelsMotionMountBleConfigEntry,
    async_add_entities: AddEntitiesCallback,
):
    """Set up the RefreshData and SelectPreset buttons."""
    coordinator: VogelsMotionMountBleCoordinator = config_entry.runtime_data.coordinator

    hass.services.async_register(
        DOMAIN,
        HA_SERVICE_SET_MULTI_PIN_FEATURES,
        _set_multi_pin_features,
    )

    async_add_entities(
        [
            MultiPinFeatureChangePresetsSwitch(coordinator),
            MultiPinFeatureChangeNameSwitch(coordinator),
            MultiPinFeatureDisableChannelSwitch(coordinator),
            MultiPinFeatureChangeTvOnOffDetectionSwitch(coordinator),
            MultiPinFeatureChangeDefaultPositionSwitch(coordinator),
            MultiPinFeatureStartCalibrationSwitch(coordinator),
        ]
    )


class MultiPinFeatureChangePresetsSwitch(VogelsMotionMountBleBaseEntity, SwitchEntity):
    """Set up the Switch to change multi pin feature change presets."""

    _attr_unique_id = "multi_pin_feature_change_presets"
    _attr_translation_key = _attr_unique_id

    @property
    def available(self) -> bool:
        """Set availability of multi pin features."""
        return (
            self.coordinator.data.pin_setting is not None and
            self.coordinator.data.pin_setting is VogelsMotionMountPinSettings.Multi
        )

    @property
    def is_on(self) -> bool:
        return self.coordinator.data.multi_pin_features.change_presets

    async def async_toggle(self, **kwargs):
        """Toggle if change presets is on or off."""
        await self.coordinator.api.set_multi_pin_features(
            change_presets=not self.is_on
        )

class MultiPinFeatureChangeNameSwitch(VogelsMotionMountBleBaseEntity, SwitchEntity):
    """Set up the Switch to change multi pin feature change name."""

    _attr_unique_id = "multi_pin_feature_change_name"
    _attr_translation_key = _attr_unique_id

    @property
    def available(self) -> bool:
        """Set availability of multi pin features."""
        return (
            self.coordinator.data.pin_setting is not None and
            self.coordinator.data.pin_setting is VogelsMotionMountPinSettings.Multi
        )

    @property
    def is_on(self) -> bool:
        return self.coordinator.data.multi_pin_features.change_name

    async def async_toggle(self, **kwargs):
        """Toggle if change name is on or off."""
        await self.coordinator.api.set_multi_pin_features(
            change_name=not self.is_on
        )

class MultiPinFeatureDisableChannelSwitch(VogelsMotionMountBleBaseEntity, SwitchEntity):
    """Set up the Switch to change multi pin feature disable channel."""

    _attr_unique_id = "multi_pin_feature_disable_channel"
    _attr_translation_key = _attr_unique_id

    @property
    def available(self) -> bool:
        """Set availability of multi pin features."""
        return (
            self.coordinator.data.pin_setting is not None and
            self.coordinator.data.pin_setting is VogelsMotionMountPinSettings.Multi
        )

    @property
    def is_on(self) -> bool:
        return self.coordinator.data.multi_pin_features.disable_channel

    async def async_toggle(self, **kwargs):
        """Toggle if disable channeld is on or off."""
        await self.coordinator.api.set_multi_pin_features(
            disable_channel=not self.is_on
        )

class MultiPinFeatureChangeTvOnOffDetectionSwitch(VogelsMotionMountBleBaseEntity, SwitchEntity):
    """Set up the Switch to change multi pin feature change tv on off detection."""

    _attr_unique_id = "multi_pin_feature_change_tv_on_off_detection"
    _attr_translation_key = _attr_unique_id

    @property
    def available(self) -> bool:
        """Set availability of multi pin features."""
        return (
            self.coordinator.data.pin_setting is not None and
            self.coordinator.data.pin_setting is VogelsMotionMountPinSettings.Multi
        )

    @property
    def is_on(self) -> bool:
        return self.coordinator.data.multi_pin_features.change_tv_on_off_detection

    async def async_toggle(self, **kwargs):
        """Toggle if change tv on off detection is on or off."""
        await self.coordinator.api.set_multi_pin_features(
            change_tv_on_off_detection=not self.is_on
        )

class MultiPinFeatureChangeDefaultPositionSwitch(VogelsMotionMountBleBaseEntity, SwitchEntity):
    """Set up the Switch to change multi pin feature change default position."""

    _attr_unique_id = "multi_pin_feature_change_default_position"
    _attr_translation_key = _attr_unique_id

    @property
    def available(self) -> bool:
        """Set availability of multi pin features."""
        return (
            self.coordinator.data.pin_setting is not None and
            self.coordinator.data.pin_setting is VogelsMotionMountPinSettings.Multi
        )

    @property
    def is_on(self) -> bool:
        return self.coordinator.data.multi_pin_features.change_default_position

    async def async_toggle(self, **kwargs):
        """Toggle if change default position is on or off."""
        await self.coordinator.api.set_multi_pin_features(
            change_default_position=not self.is_on
        )

class MultiPinFeatureStartCalibrationSwitch(VogelsMotionMountBleBaseEntity, SwitchEntity):
    """Set up the Switch to change multi pin feature start calibration."""

    _attr_unique_id = "multi_pin_feature_start_calibration"
    _attr_translation_key = _attr_unique_id

    @property
    def available(self) -> bool:
        """Set availability of multi pin features."""
        return (
            self.coordinator.data.pin_setting is not None and
            self.coordinator.data.pin_setting is VogelsMotionMountPinSettings.Multi
        )

    @property
    def is_on(self) -> bool:
        return self.coordinator.data.multi_pin_features.start_calibration

    async def async_toggle(self, **kwargs):
        """Toggle if start calibration is on or off."""
        await self.coordinator.api.set_multi_pin_features(
            start_calibration=not self.is_on
        )