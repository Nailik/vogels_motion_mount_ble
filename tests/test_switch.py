import pytest
from syrupy.assertion import SnapshotAssertion
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er
from unittest.mock import patch

from . import setup_integration

from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
    snapshot_platform,
)

from custom_components.vogels_motion_mount_ble.switch import (
    MultiPinFeatureChangePresetsSwitch,
    MultiPinFeatureChangeNameSwitch,
    MultiPinFeatureDisableChannelSwitch,
    MultiPinFeatureChangeTvOnOffDetectionSwitch,
    MultiPinFeatureChangeDefaultPositionSwitch,
    MultiPinFeatureStartCalibrationSwitch,
)
from unittest.mock import AsyncMock
from dataclasses import replace


# -------------------------------
# endregion
# region Setup
# -------------------------------


async def test_all_entities(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test all entities."""
    with patch(
        "custom_components.vogels_motion_mount_ble.PLATFORMS", [Platform.SWITCH]
    ):
        await setup_integration(hass, mock_config_entry)

    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)


# -------------------------------
# endregion
# region Actions
# -------------------------------


@pytest.mark.asyncio
@pytest.mark.parametrize(
    "switch_cls, field",
    [
        (MultiPinFeatureChangePresetsSwitch, "change_presets"),
        (MultiPinFeatureChangeNameSwitch, "change_name"),
        (MultiPinFeatureDisableChannelSwitch, "disable_channel"),
        (MultiPinFeatureChangeTvOnOffDetectionSwitch, "change_tv_on_off_detection"),
        (MultiPinFeatureChangeDefaultPositionSwitch, "change_default_position"),
        (MultiPinFeatureStartCalibrationSwitch, "start_calibration"),
    ],
)
async def test_switch_toggle_actions(mock_coord, switch_cls, field):
    # Ensure initial field is False
    features = replace(mock_coord.data.multi_pin_features, **{field: False})
    mock_coord.data.multi_pin_features = features
    mock_coord.set_multi_pin_features = AsyncMock()

    entity = switch_cls(mock_coord)
    assert entity.is_on is False

    # Turn on → should toggle field True
    await entity.async_turn_on()
    called_features = mock_coord.set_multi_pin_features.await_args[0][0]
    assert getattr(called_features, field) is True

    mock_coord.set_multi_pin_features.reset_mock()

    # Update coordinator state → set field True
    mock_coord.data.multi_pin_features = replace(features, **{field: True})
    assert entity.is_on is True

    # Turn off → should toggle field False
    await entity.async_turn_off()
    called_features = mock_coord.set_multi_pin_features.await_args[0][0]
    assert getattr(called_features, field) is False


# -------------------------------
# endregion
# -------------------------------
