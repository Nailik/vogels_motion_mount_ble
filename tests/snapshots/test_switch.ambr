# serializer version: 1
# name: test_all_entities[switch.mount_change_default_position-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'switch.mount_change_default_position',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:security',
    'original_name': 'Change default position',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'change_default_position',
    'unique_id': 'change_default_position',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[switch.mount_change_default_position-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Change default position',
      'icon': 'mdi:security',
    }),
    'context': <ANY>,
    'entity_id': 'switch.mount_change_default_position',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_all_entities[switch.mount_change_name-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'switch.mount_change_name',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:security',
    'original_name': 'Change name',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'change_name',
    'unique_id': 'change_name',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[switch.mount_change_name-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Change name',
      'icon': 'mdi:security',
    }),
    'context': <ANY>,
    'entity_id': 'switch.mount_change_name',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_all_entities[switch.mount_change_presets-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'switch.mount_change_presets',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:security',
    'original_name': 'Change presets',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'change_presets',
    'unique_id': 'change_presets',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[switch.mount_change_presets-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Change presets',
      'icon': 'mdi:security',
    }),
    'context': <ANY>,
    'entity_id': 'switch.mount_change_presets',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_all_entities[switch.mount_change_tv_on_off_detection-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'switch.mount_change_tv_on_off_detection',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:security',
    'original_name': 'Change tv on off detection',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'change_tv_on_off_detection',
    'unique_id': 'change_tv_on_off_detection',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[switch.mount_change_tv_on_off_detection-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Change tv on off detection',
      'icon': 'mdi:security',
    }),
    'context': <ANY>,
    'entity_id': 'switch.mount_change_tv_on_off_detection',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_all_entities[switch.mount_disable_channel-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'switch.mount_disable_channel',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:security',
    'original_name': 'Disable channel',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'disable_channel',
    'unique_id': 'disable_channel',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[switch.mount_disable_channel-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Disable channel',
      'icon': 'mdi:security',
    }),
    'context': <ANY>,
    'entity_id': 'switch.mount_disable_channel',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_all_entities[switch.mount_start_calibration-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': <EntityCategory.CONFIG: 'config'>,
    'entity_id': 'switch.mount_start_calibration',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': 'mdi:security',
    'original_name': 'Start calibration',
    'platform': 'vogels_motion_mount_ble',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': 0,
    'translation_key': 'start_calibration',
    'unique_id': 'start_calibration',
    'unit_of_measurement': None,
  })
# ---
# name: test_all_entities[switch.mount_start_calibration-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Mount Start calibration',
      'icon': 'mdi:security',
    }),
    'context': <ANY>,
    'entity_id': 'switch.mount_start_calibration',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
